drivers/sensor/stm32_vref/stm32_vref.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2023 Kenneth J. Miller <ken@miller.ec>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/adc.h>
+#include <zephyr/logging/log.h>
+#if defined(CONFIG_SOC_SERIES_STM32H5X)
+#include <stm32_ll_icache.h>
+#endif /* CONFIG_SOC_SERIES_STM32H5X */
+
+LOG_MODULE_REGISTER(stm32_vref, CONFIG_SENSOR_LOG_LEVEL);
+#if DT_HAS_COMPAT_STATUS_OKAY(st_stm32_vref)
+#define DT_DRV_COMPAT st_stm32_vref
+#else
+#error "No compatible devicetree node found"
+#endif
+
+struct stm32_vref_data {
+	const struct device *adc;
+	const struct adc_channel_cfg adc_cfg;
+	struct adc_sequence adc_seq;
+	struct k_mutex mutex;
+	int16_t sample_buffer;
+	int16_t raw; /* raw adc Sensor value */
+};
+
+struct stm32_vref_config {
+	uint16_t *cal_addr;
+	int cal_mv;
+};
+
+static int stm32_vref_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	struct stm32_vref_data *data = dev->data;
+	struct adc_sequence *sp = &data->adc_seq;
+	int rc;
+
+	if (chan != SENSOR_CHAN_ALL && chan != SENSOR_CHAN_VOLTAGE) {
+		return -ENOTSUP;
+	}
+
+	k_mutex_lock(&data->mutex, K_FOREVER);
+
+	rc = adc_channel_setup(data->adc, &data->adc_cfg);
+	if (rc) {
+		LOG_DBG("Setup AIN%u got %d", data->adc_cfg.channel_id, rc);
+		goto unlock;
+	}
+
+	rc = adc_read(data->adc, sp);
+	if (rc == 0) {
+		data->raw = data->sample_buffer;
+	}
+
+unlock:
+	k_mutex_unlock(&data->mutex);
+
+	return rc;
+}
+
+static int stm32_vref_channel_get(const struct device *dev, enum sensor_channel chan,
+				  struct sensor_value *val)
+{
+	struct stm32_vref_data *data = dev->data;
+	const struct stm32_vref_config *cfg = dev->config;
+	float vref;
+
+	if (chan != SENSOR_CHAN_VOLTAGE) {
+		return -ENOTSUP;
+	}
+
+/*
+ * ERRATA: STM32H5X: bus fault errors occur when reading engineering bytes with
+ * icache enabled.
+ * See https://github.com/zephyrproject-rtos/zephyr/commit/065a8f2
+ */
+#if defined(CONFIG_SOC_SERIES_STM32H5X)
+	LL_ICACHE_Disable();
+#endif /* CONFIG_SOC_SERIES_STM32H5X */
+
+	/* Calculate VREF+ using VREFINT bandgap voltage and calibration data */
+	vref = cfg->cal_mv * (*cfg->cal_addr) / data->raw;
```

Comment:
You may want to guard against dividing by zero

Reply:
Good call, will amend the commit to add this. Thanks!